server/sonar-server/src/main/java/org/sonar/server/computation/step/PersistCrossProjectDuplicationIndexStep.java

```diff
@@ -0,0 +1,114 @@
+import org.sonar.server.computation.component.Component;
+import org.sonar.server.computation.component.CrawlerDepthLimit;
+import org.sonar.server.computation.component.DbIdsRepository;
+import org.sonar.server.computation.component.DepthTraversalTypeAwareCrawler;
+import org.sonar.server.computation.component.TreeRootHolder;
+import org.sonar.server.computation.component.TypeAwareVisitorAdapter;
+
+import static org.sonar.server.computation.component.ComponentVisitor.Order.PRE_ORDER;
+
+/**
+ * Persist cross project duplications text blocks into DUPLICATIONS_INDEX table
+ */
+public class PersistCrossProjectDuplicationIndexStep implements ComputationStep {
+
+  private final DbClient dbClient;
+  private final TreeRootHolder treeRootHolder;
+  private final BatchReportReader reportReader;
+  private final AnalysisMetadataHolder analysisMetadataHolder;
+  private final DbIdsRepository dbIdsRepository;
+
+  public PersistCrossProjectDuplicationIndexStep(DbClient dbClient, DbIdsRepository dbIdsRepository, TreeRootHolder treeRootHolder, BatchReportReader reportReader,
+    AnalysisMetadataHolder analysisMetadataHolder) {
+    this.dbClient = dbClient;
+    this.treeRootHolder = treeRootHolder;
+    this.reportReader = reportReader;
+    this.analysisMetadataHolder = analysisMetadataHolder;
+    this.dbIdsRepository = dbIdsRepository;
+  }
+
+  @Override
+  public void execute() {
+    DbSession session = dbClient.openSession(true);
+    try {
+      if (analysisMetadataHolder.isCrossProjectDuplicationEnabled() && analysisMetadataHolder.getBranch() == null) {
+        Component project = treeRootHolder.getRoot();
+        long projectSnapshotId = dbIdsRepository.getSnapshotId(project);
+        new DepthTraversalTypeAwareCrawler(new DuplicationVisitor(session, projectSnapshotId)).visit(project);
+      }
+      session.commit();
+    } finally {
+      MyBatis.closeQuietly(session);
+    }
+  }
+
+  private class DuplicationVisitor extends TypeAwareVisitorAdapter {
+
+    private final DbSession session;
+    private final long projectSnapshotId;
+
+    private DuplicationVisitor(DbSession session, long projectSnapshotId) {
+      super(CrawlerDepthLimit.FILE, PRE_ORDER);
+      this.session = session;
+      this.projectSnapshotId = projectSnapshotId;
+    }
+
+    @Override
+    public void visitFile(Component file) {
+      visitComponent(file);
+    }
+
+    private void visitComponent(Component component) {
+      int indexInFile = 0;
+      try (CloseableIterator<BatchReport.CpdTextBlock> blocks = reportReader.readCpdTextBlocks(component.getReportAttributes().getRef())) {
+        while (blocks.hasNext()) {
+          BatchReport.CpdTextBlock block = blocks.next();
+          dbClient.duplicationDao().insert(session, new DuplicationUnitDto()
+            .setHash(block.getHash())
+            .setStartLine(block.getStartLine())
+            .setEndLine(block.getEndLine())
+            .setIndexInFile(indexInFile++)
```

Comment:
![MAJOR](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/severity-major.png) Extract this increment or decrement operator into a dedicated statement. [![rule](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/rule.png)](https://dory.sonarsource.com/coding_rules#rule_key=squid%3AS881)